.pre-commit-config.yaml

```
@@ -98,3 +104,15 @@ repos:
     types:
     - c++
     entry: ./tools/check-style.sh
+
+- repo: local
+  hooks:
+  - id: docker-clang-format
+    name: Docker Clang Format
+    language: docker_image
+    types:
+    - c++
+    entry: unibeautify/clang-format:latest
+    args:
+    - -style=file
+    - -i
```

It is not necessary to run clang-format within a docker container. In [MoveIt we are using it directly](https://github.com/ros-planning/moveit/blob/master/.pre-commit-config.yaml#L35-L43) like this:
```suggestion

- repo: local
  hooks:
      - id: clang-format
        name: clang-format
        description: Format files with clang-format.
        entry: clang-format-10
        language: system
        files: \.(c|cc|cxx|cpp|frag|glsl|h|hpp|hxx|ih|ispc|ipp|java|js|m|proto|vert)$
        args: ['-fallback-style=none', '-i']
```
This just requires installing `clang-format` before running the `pre-commit` action:
https://github.com/ros-planning/moveit/blob/master/.github/workflows/format.yml#L18-L19

Note, that the different versions of `clang-format` are not interoperable. You need to choose (and stick to) a particular version.

---

The docker solution a) works everywhere Docker does, b) doesn't require special pre-installs locally (which is one of the main advantages of pre-commit), and c) ensures the same version is used everywhere, rather than depending on what a user happened to pre-install.

Now, what you _can_ do (and I do have for https://github.com/cliutils/cli11, for example) is two pre-commit files. If you can't run docker locally but don't mind installing the correct version of clang-format, you can specifically ask for the "no-docker" version. But it's a bit of a pain to keep them in sync, and most people just use the docker one because it's simpler.

---

So, you then manually configure your local pre-commit git-hook to use the `no-docker` version?

---

Yes. Well, I don't, I use docker, but I had one person request this. (I think they have since moved to docker)